charmcraft/application/commands/store.py

```diff
@@ -830,6 +833,141 @@ def run(self, parsed_args):
+
+        # Check snapcraft for equiv logic
+        from_channel = charmcraft.store.models.ChannelData.from_str(
+            parsed_args.from_channel
+        )
+        to_channel = charmcraft.store.models.ChannelData.from_str(
+            parsed_args.to_channel
+        )
+        if None in (from_channel.track, to_channel.track):
+            package_metadata = store.get_package_metadata(name)
+            default_track = package_metadata.default_track
+            if from_channel.track is None:
+                from_channel = dataclasses.replace(from_channel, track=default_track)
+            if to_channel.track is None:
+                to_channel = dataclasses.replace(to_channel, track=default_track)
+
+        if to_channel == from_channel:
+            raise CraftError(
+                "Cannot promote from a channel to the same channel.",
+                retcode=64,  # Replace with os.EX_USAGE once we drop Windows.
+            )
+        if to_channel.risk > from_channel.risk:
+            command_parts = [
+                self._app.name,
+                f"--from-channel={to_channel.name}",
+                f"--to-channel={from_channel.name}",
+                self.name,
+            ]
+            command = " ".join(command_parts)
+            raise CraftError(
+                f"Target channel ({to_channel.name}) must be lower risk "
+                f"than the source channel ({from_channel.name}).",
+                resolution=f"Did you mean: {command}",
+            )
+        if to_channel.track != from_channel.track:
+            if not parsed_args.yes and not utils.confirm_with_user(
+                "Did you mean to promote to a different track? (from "
+                f"{from_channel.track} to {to_channel.track})",
+            ):
+                emit.message("Cancelling.")
+                return 64  # Replace with os.EX_USAGE once we drop Windows.
+
+        candidates = store.get_revisions_on_channel(name, from_channel.name)
+
+        def get_base_strings(bases):
+            if bases is None:
+                return ""
+            return ",".join(
+                f"{base.name}@{base.channel}:{base.architecture}" for base in bases
+            )
+
+        presentable_candidates = [
+            {
+                "Revision": info["revision"],
+                "Platforms": get_base_strings(info["bases"]),
+                "Resource revisions": ", ".join(
+                    f"{res['name']}: {res['revision']}" for res in info["resources"]
+                ),
+            }
+            for info in sorted(candidates, key=lambda x: x["revision"])
+        ]
+        emit.progress(
+            f"The following revisions are on the {from_channel.name} channel:",
+            permanent=True,
+        )
+        with emit.pause():
+            print(
+                tabulate(presentable_candidates, tablefmt="plain", headers="keys"),
+                file=sys.stderr,
+            )
```

Comment:
Any particular reason why it can't be a progress as well (beside not logging it) ?

Reply:
`emit.progress` tries to represent it as a single line, even with `permanent=True`.

Reply:
Would it work this way ? It seems to be a bit more consistent and does not leak the `sys.stderr` implementation detail
```suggestion
        with emit.open_stream() as stream:
            print(
                tabulate(presentable_candidates, tablefmt="plain", headers="keys"),
                file=stream,
            )
```

Reply:
`open_stream()` doesn't result in a file-like object, just an integer with the file pointer, so you'd have to then open that, and it gets ugly.

Reply:
You can probably do `.os.write()` then instead of print, but it's up to you.

I did emit.progress() previously on multi line output and it worked, that's why I asked in the first place.

Reply:
I used `emit.progress` initially and it didn't work. I don't want to use `os.write` here because it requires handling a bunch of things that `print` already handles, so if we really wanted to use the stream I'd probably use `open` to create a file object from that descriptor and then `print` to that.